blockchain/state_manager/src/vm_circ_supply.rs

```diff
@@ -0,0 +1,284 @@
+    Ok(power_locked + market_locked)
+}
+
+pub fn get_fil_burnt<DB: BlockStore>(
+    state_tree: &StateTree<DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let burnt_actor = get_actor_state(state_tree, &*BURNT_FUNDS_ACTOR_ADDR)?;
+
+    Ok(burnt_actor.balance)
+}
+
+pub fn get_circulating_supply<'a, DB: BlockStore>(
+    pre_ignition: &GenesisInfo,
+    post_ignition: &GenesisInfo,
+    height: ChainEpoch,
+    state_tree: &StateTree<'a, DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let fil_vested = get_fil_vested(&pre_ignition, &post_ignition, height, &state_tree)?;
+    let fil_mined = get_fil_mined(&state_tree)?;
+    let fil_burnt = get_fil_burnt(&state_tree)?;
+    let fil_locked = get_fil_locked(&state_tree)?;
+    let fil_reserve_distributed = if height > UPGRADE_ACTORS_V2_HEIGHT {
+        get_fil_reserve_disbursed(&state_tree)?
+    } else {
+        TokenAmount::default()
+    };
+    let fil_circulating = BigInt::max(
+        &fil_vested + &fil_mined + fil_reserve_distributed - &fil_burnt - &fil_locked,
+        TokenAmount::default(),
+    );
+
+    Ok(fil_circulating)
+}
+
+fn init_genesis_info<DB: BlockStore>(bs: &DB) -> Result<GenesisInfo, Box<dyn StdError>> {
+    let mut ignition = GenesisInfo::default();
+
+    let genesis_block = genesis(bs)?.ok_or_else(|| "Genesis Block doesnt exist".to_string())?;
+
+    let gts = Tipset::new(vec![genesis_block])?;
+
+    // Parent state of genesis tipset is tipset state
+    let st = gts.parent_state();
+
+    let state_tree = StateTree::new_from_root(bs, &st)?;
+
+    ignition.genesis_market_funds = get_fil_market_locked(&state_tree)?;
+    ignition.genesis_pledge = get_fil_power_locked(&state_tree)?;
+
+    Ok(ignition)
+}
+
+pub fn setup_preignition_genesis_actors_testnet<DB: BlockStore>(
+    bs: &DB,
+) -> Result<GenesisInfo, Box<dyn StdError>> {
+    let mut pre_ignition = init_genesis_info(bs)?;
+
+    for (unlock_duration, initial_balance) in &*TOTALS_BY_EPOCH {
+        let ms = multisig::State {
+            signers: vec![],
+            num_approvals_threshold: 0,
+            next_tx_id: multisig::TxnID(0),
+            initial_balance: initial_balance.clone(),
+            start_epoch: ChainEpoch::default(),
+            unlock_duration: *unlock_duration,
+            // Default Cid is ok here because this field is never read
+            pending_txs: Cid::default(),
+        };
+        pre_ignition.genesis_msigs.push(ms);
+    }
+
+    Ok(pre_ignition)
+}
+
+pub fn setup_postignition_genesis_actors_testnet<DB: BlockStore>(
+    bs: &DB,
+) -> Result<GenesisInfo, Box<dyn StdError>> {
+    let mut post_ignition = init_genesis_info(bs)?;
+
+    for (unlock_duration, initial_balance) in &*TOTALS_BY_EPOCH {
+        let ms = multisig::State {
+            signers: vec![],
+            num_approvals_threshold: 0,
+            next_tx_id: multisig::TxnID(0),
+            initial_balance: initial_balance * FILECOIN_PRECISION,
```

Comment:
i'd add a note here similar to how to lotus has to indicate why you multiply by the precision here but not in preignition